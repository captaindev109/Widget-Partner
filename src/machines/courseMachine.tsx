import { createMachine } from 'xstate';

export const MAIN = {
  TRY_HEDVIG: 'TRY_HEDVIG',
  ARE_YOU_INSURED: 'ARE_YOU_INSURED',
  YOUR_INFORMATION: 'YOUR_INFORMATION',
  GET_AN_OFFER: 'GET_AN_OFFER',
  CHECKOUT: 'CHECKOUT',
  CONFIRMATION: 'CONFIRMATION',
};

type MainServiceEvents = { type: 'PREVIOUS' } | { type: 'NEXT' };

interface MainServiceStates {
  states: {
    TRY_HEDVIG: Record<string, string>;
    ARE_YOU_INSURED: Record<string, string>;
    YOUR_INFORMATION: Record<string, string>;
    GET_AN_OFFER: Record<string, string>;
    CHECKOUT: Record<string, string>;
    CONFIRMATION: Record<string, string>;
  };
  value: string;
  context: Record<string, string>;
}

interface MainServiceContext {
  currentState?: string;
}

export const mainMachine = createMachine<MainServiceContext, MainServiceEvents, MainServiceStates>({
  // Machine identifier
  id: 'mainMachine',

  // Initial state
  initial: MAIN.TRY_HEDVIG,

  // Context
  context: {
    currentState: MAIN.TRY_HEDVIG,
  },

  // State definitions
  states: {
    TRY_HEDVIG: {
      on: {
        NEXT: MAIN.ARE_YOU_INSURED,
      },
    },
    ARE_YOU_INSURED: {
      on: {
        NEXT: MAIN.YOUR_INFORMATION,
        PREVIOUS: MAIN.TRY_HEDVIG,
      },
    },
    YOUR_INFORMATION: {
      on: {
        NEXT: MAIN.GET_AN_OFFER,
        PREVIOUS: MAIN.ARE_YOU_INSURED,
      },
    },
    GET_AN_OFFER: {
      on: {
        NEXT: MAIN.CHECKOUT,
        PREVIOUS: MAIN.YOUR_INFORMATION,
      },
    },
    CHECKOUT: {
      on: {
        NEXT: MAIN.CONFIRMATION,
        PREVIOUS: MAIN.GET_AN_OFFER,
      },
    },
    CONFIRMATION: {
      on: {
        PREVIOUS: MAIN.CHECKOUT,
      },
    },
  },
});
