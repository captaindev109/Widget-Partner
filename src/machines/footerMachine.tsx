import { createMachine, assign } from 'xstate';

export const FOOTER_COURSES = {
  PRIMARY: 'PRIMARY',
  SECONDARY: 'SECONDARY',
};

export type FooterServiceEvents = { type: 'PRIMARY' } | { type: 'SECONDARY' } | { type: 'HEIGHT'; value: number };

export interface FooterServiceStates {
  states: {
    PRIMARY: Record<string, string>;
    SECONDARY: Record<string, string>;
  };
  value: string;
  context: FooterServiceContext;
}

export interface FooterServiceContext {
  currentState?: string;
  height: number;
}
const update_height = (context: FooterServiceContext, event: FooterServiceEvents) => {
  return event.type === 'HEIGHT' ? event.value : 0;
};

const footerMachine = createMachine<FooterServiceContext, FooterServiceEvents, FooterServiceStates>({
  id: 'footer_machine',
  initial: FOOTER_COURSES.PRIMARY,
  context: {
    currentState: FOOTER_COURSES.PRIMARY,
    height: 0,
  },
  states: {
    PRIMARY: {
      on: {
        SECONDARY: FOOTER_COURSES.SECONDARY,
      },
    },
    SECONDARY: {
      on: {
        PRIMARY: FOOTER_COURSES.PRIMARY,
      },
    },
  },
  on: {
    HEIGHT: {
      actions: assign({ height: update_height }),
    },
  },
});

export default footerMachine;
