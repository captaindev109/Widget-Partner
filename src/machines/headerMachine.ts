import { createMachine } from 'xstate';

export const HEADER_STATES = {
  PRIMARY: 'PRIMARY',
  SECONDARY: 'SECONDARY',
  TERTIARY: 'TERTIARY',
};

export type HeaderServiceEvents = { type: 'PRIMARY' } | { type: 'SECONDARY' } | { type: 'TERTIARY' };

export interface HeaderServiceStates {
  states: {
    PRIMARY: Record<string, string>;
    SECONDARY: Record<string, string>;
    TERTIARY: Record<string, string>;
  };
  value: string;
  context: Record<string, string>;
}

export interface HeaderServiceContext {
  currentState?: string;
}

const headerMachine = createMachine<HeaderServiceContext, HeaderServiceEvents, HeaderServiceStates>({
  id: 'header_machine',
  initial: HEADER_STATES.PRIMARY,
  context: {
    currentState: HEADER_STATES.PRIMARY,
  },
  states: {
    PRIMARY: {
      on: {
        SECONDARY: HEADER_STATES.SECONDARY,
        TERTIARY: HEADER_STATES.TERTIARY,
      },
    },
    SECONDARY: {
      on: {
        PRIMARY: HEADER_STATES.PRIMARY,
        TERTIARY: HEADER_STATES.TERTIARY,
      },
    },
    TERTIARY: {
      on: {
        PRIMARY: HEADER_STATES.PRIMARY,
        SECONDARY: HEADER_STATES.SECONDARY,
        TERTIARY: HEADER_STATES.TERTIARY,
      },
    },
  },
});

export default headerMachine;
