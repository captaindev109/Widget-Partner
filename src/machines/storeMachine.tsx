import {
  BasicAny,
  BasicString,
  ExoticBundleVariation,
  ExoticQuote,
  StoreCart,
  StoreCartItem,
  StoreCheckout,
  StorePartner,
  StoreUser,
  StoreUserInput,
} from '@/types/types';

import { assign, createMachine } from 'xstate';

type StoreEvents = {
  value?: BasicAny;
  type:
    | 'UPDATE_ROUTES'
    | 'UPDATE_CURRENT_ROUTE'
    | 'UPDATE_BUNDLE_COST'
    | 'UPDATE_BUNDLE_VARIATIONS'
    | 'UPDATE_BUNDLE_QUOTES'
    | 'UPDATE_FORM_DATA'
    | 'UPDATE_SELECTED_BUNDLE_TAB'
    | 'CREATE_CART'
    | 'UPDATE_CART_ITEMS'
    | 'UPDATE_USER'
    | 'UPDATE_START_DATE'
    | 'UPDATE_CHECKOUT'
    | 'CREATE_PARTNER';
};

export interface ContextType {
  _routes: BasicAny[];
  _current_route: BasicString;
  _store: BasicAny;
  _bundleVariations: ExoticBundleVariation[];
  _quotes: ExoticQuote[];
  _formData: BasicAny;
  _selectedBundleTab: BasicString;
  _cart: StoreCart;
  _user: StoreUser;
  _checkout: StoreCheckout;
  _partner: StorePartner;
}

export interface StoreStates {
  states: {
    loading: object;
  };
  value: string;
  context: ContextType;
}

type StoreContext = BasicAny;

const update_routes = (context: ContextType, event: BasicAny) => {
  return event.routes;
};

const update_current_route = (context: ContextType, event: BasicAny) => {
  return event.current_route;
};

const update_bundle_variations = (context: ContextType, event: BasicAny) => {
  return event.value;
};

const update_bundle_quotes = (context: ContextType, event: BasicAny) => {
  return event.value;
};

const update_form_data = (context: ContextType, event: BasicAny) => {
  const { _formData } = context;
  const { value } = event;

  const newFormDataObject = { ..._formData, ...value };
  return newFormDataObject;
};

const update_selected_bundle = (context: ContextType, event: BasicAny) => {
  console.log('update_selected_bundle', event);
  return event.value;
};

const create_cart = (context: ContextType, event: { value: string }): BasicAny => {
  console.log('create_cart', event);
  const { _cart } = context;
  const { value } = event;
  const cart = { ..._cart, id: value };
  return cart;
};

const create_partner = (context: ContextType, event: { value: string }): BasicAny => {
  console.log('create_partner', event);
  return event.value;
};

const update_cart_items = (context: ContextType, event: { value: StoreCartItem[] }): BasicAny => {
  console.log('update_cart_items', event);
  const { _cart } = context;
  const { value } = event;

  const cart = { ..._cart, items: value };
  return cart;
};

const update_user = (context: ContextType, event: { value: StoreUserInput }): BasicAny => {
  console.log('update_user', event);
  const { _user } = context;
  const { value } = event;
  const newUser = { ..._user, ...value };
  return newUser;
};

const update_quote_start_date = (context: ContextType, event: { value: { id: string; date: Date } }): BasicAny => {
  console.log('update_quote_start_date', event);
  const { value } = event;
  const { id, date } = value;
  const { _quotes } = context;
  const newQuotes = _quotes.map((quote) => {
    if (quote.id === id) {
      return { ...quote, startDate: date };
    }
    return quote;
  });
  return newQuotes;
};

const update_checkout = (context: ContextType, event: { value: StoreCheckout }): BasicAny => {
  console.log('update_checkout', event);
  return event.value;
};

export const storeMachine = createMachine<StoreContext, StoreEvents, StoreStates>({
  // Machine identifier
  id: 'store',

  // Initial state
  initial: 'loading',

  // Context
  context: {
    _routes: [],
    _current_route: '',
    _bundleVariations: [],
    _quotes: [],
    _formData: {},
    _selectedBundleTab: '',
    _cart: {
      id: '',
      items: [],
      cost: {
        amount: '',
        currency: '',
      },
    },
    _user: {
      firstName: '',
      lastName: '',
      birthDate: '',
      ssn: '',
      email: '',
      phoneNumber: '',
    },
    _partner: {
      name: 'AVY',
    },
    _checkout: null,
  },

  // State definitions
  states: {
    loading: {
      on: {
        UPDATE_ROUTES: {
          actions: assign({ _routes: update_routes }),
        },
        UPDATE_CURRENT_ROUTE: {
          actions: assign({ _current_route: update_current_route }),
        },
        UPDATE_BUNDLE_VARIATIONS: {
          actions: assign({ _bundleVariations: update_bundle_variations }),
        },
        UPDATE_BUNDLE_QUOTES: {
          actions: assign({ _quotes: update_bundle_quotes }),
        },
        UPDATE_FORM_DATA: {
          actions: assign({ _formData: update_form_data }),
        },
        UPDATE_SELECTED_BUNDLE_TAB: {
          actions: assign({ _selectedBundleTab: update_selected_bundle }),
        },
        CREATE_CART: {
          actions: assign({ _cart: create_cart }),
        },
        UPDATE_CART_ITEMS: {
          actions: assign({ _cart: update_cart_items }),
        },
        UPDATE_USER: {
          actions: assign({ _user: update_user }),
        },
        UPDATE_START_DATE: {
          actions: assign({ _quotes: update_quote_start_date }),
        },
        UPDATE_CHECKOUT: {
          actions: assign({ _checkout: update_checkout }),
        },
        CREATE_PARTNER: {
          actions: assign({ _partner: create_partner }),
        },
      },
    },
  },
});
