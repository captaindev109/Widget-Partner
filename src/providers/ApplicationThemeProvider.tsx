import { ThemeProvider, DefaultTheme } from 'styled-components';
import GlobalStyles from '@/styles/Global';
import React from 'react';
import { ChildrenReturnType } from '@/types/types';

interface ApplicationThemeProviderProps {
  children?: ChildrenReturnType;
  theme: DefaultTheme;
}

const ApplicationThemeProvider = ({ children, theme }: ApplicationThemeProviderProps) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      {children}
    </ThemeProvider>
  );
};

export default ApplicationThemeProvider;
