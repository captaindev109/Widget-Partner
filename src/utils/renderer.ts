import React from 'react';
import { ConfigProps, Index, EachComponentObjectChildren, BasicAny } from '@/types/types';
import { UIComponents } from './components';

const recursivelyCreateReactComponents = (
  componentToCreate: BasicAny,
  configuration: ConfigProps,
  children: BasicAny
) => {
  if (typeof componentToCreate !== 'undefined') {
    return React.createElement(
      componentToCreate,
      configuration,
      children &&
        (typeof children === 'string'
          ? children
          : children.map((component: EachComponentObjectChildren, index: Index) => viewRenderer(component, index)))
    );
  }
  null;
};

const viewRenderer: BasicAny = (config: ConfigProps, index: Index) => {
  const children = config.children;
  const newConfig = { ...config, key: index };
  const LookForComponentInUILibrary = UIComponents[config.component];
  return recursivelyCreateReactComponents(LookForComponentInUILibrary, newConfig, children);
};

export default viewRenderer;
