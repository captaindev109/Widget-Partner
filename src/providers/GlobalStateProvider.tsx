import { mainMachine } from '@/machines/courseMachine';
import headerMachine from '@/machines/headerMachine';
import footerMachine from '@/machines/footerMachine';
import React, { createContext } from 'react';
import { useInterpret } from '@xstate/react';
import { ActorRefFrom } from 'xstate';
import { storeMachine } from '@/machines/storeMachine';

interface GlobalStateContextProps {
  mainService: ActorRefFrom<typeof mainMachine>;
  headerService: ActorRefFrom<typeof headerMachine>;
  footerService: ActorRefFrom<typeof footerMachine>;
  storeService: ActorRefFrom<typeof storeMachine>;
}

export const GlobalStateContext = createContext({} as GlobalStateContextProps);

interface GlobalStateProviderProps {
  children?: React.ReactNode;
}

const useDevTool = process.env.NODE_ENV === 'development';

export const GlobalStateProvider = ({ children }: GlobalStateProviderProps) => {
  const services = {
    mainService: useInterpret(mainMachine, {
      devTools: useDevTool,
    }),
    headerService: useInterpret(headerMachine, {
      devTools: useDevTool,
    }),
    footerService: useInterpret(footerMachine, {
      devTools: useDevTool,
    }),
    storeService: useInterpret(storeMachine, {
      devTools: useDevTool,
    }),
  };

  return <GlobalStateContext.Provider value={services}>{children}</GlobalStateContext.Provider>;
};
