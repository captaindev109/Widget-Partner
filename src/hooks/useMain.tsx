import { GlobalStateContext } from '@/providers/GlobalStateProvider';
import { useActor } from '@xstate/react';
import { useCallback, useContext } from 'react';
import { step } from '@/utils/routing';

const useMain = () => {
  const { mainService } = useContext(GlobalStateContext);
  const [state, send] = useActor(mainService);

  const previousCourse = useCallback(() => {
    send('PREVIOUS');
  }, [send]);

  const nextCourse = useCallback(() => {
    send('NEXT');
  }, [send]);

  return { state: state.value, nextCourse, previousCourse, step };
};

export default useMain;
