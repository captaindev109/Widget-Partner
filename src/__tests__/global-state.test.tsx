import { render, RenderResult, fireEvent } from '@testing-library/react';
import queries from '@testing-library/dom/types/queries';
import renderer from 'react-test-renderer';
import { GlobalStateProvider } from '@/providers/GlobalStateProvider';
import { FC } from 'react';
import useMain from '@/hooks/useMain';
import { MAIN } from '@/machines/courseMachine';

const GlobalStateComponent: FC = () => (
  <GlobalStateProvider>
    <GlobalStateComponentWithButtons />
  </GlobalStateProvider>
);

const GlobalStateComponentWithButtons: FC = () => {
  const { state, nextCourse, previousCourse } = useMain();
  return (
    <>
      <div data-testid="state">{state}</div>
      <button data-testid="forward" onClick={nextCourse}>
        forward
      </button>
      <button data-testid="backward" onClick={previousCourse}>
        backward
      </button>
    </>
  );
};

// Render tests
describe('Render tests', () => {
  test('Global state provider renders', () => {
    render(<GlobalStateComponent />);
  });
});

// Routing tests
describe('Routing tests', () => {
  let renderedComponent: RenderResult<typeof queries, HTMLElement>;

  beforeEach(() => {
    renderedComponent = render(<GlobalStateComponent />);
  });

  test('State exists', () => {
    const { getByTestId } = renderedComponent;

    const state = getByTestId('state');
    expect(state).toBeInTheDocument();
  });

  test('Navigate forward', () => {
    const { getByTestId } = renderedComponent;
    const state = getByTestId('state');
    const forward = getByTestId('forward');

    // Base case is TRY_HEDVIG
    expect(state).toHaveTextContent(MAIN.TRY_HEDVIG);
    fireEvent.click(forward);

    // Second case is ARE_YOU_INSURED
    expect(state).toHaveTextContent(MAIN.ARE_YOU_INSURED);

    // See so CONFIRMATION is the absolute last case
    fireEvent.click(forward);
    fireEvent.click(forward);
    fireEvent.click(forward);
    fireEvent.click(forward);
    fireEvent.click(forward);
    fireEvent.click(forward);
    expect(state).toHaveTextContent(MAIN.CONFIRMATION);
  });

  test('Navigate backward', () => {
    const { getByTestId } = renderedComponent;
    const state = getByTestId('state');
    const forward = getByTestId('forward');
    const backward = getByTestId('backward');

    // Base case is TRY_HEDVIG
    expect(state).toHaveTextContent(MAIN.TRY_HEDVIG);

    // Second case is ARE_YOU_INSURED
    fireEvent.click(forward);
    expect(state).toHaveTextContent(MAIN.ARE_YOU_INSURED);

    // Try to go back
    fireEvent.click(backward);
    expect(state).toHaveTextContent(MAIN.TRY_HEDVIG);

    // See so TRY_HEDVIG is the absolute first
    fireEvent.click(backward);
    fireEvent.click(backward);
    fireEvent.click(backward);
    expect(state).toHaveTextContent(MAIN.TRY_HEDVIG);
  });
});

// Snapshot test
describe('Snapshots', () => {
  it('Snapshot test should pass', () => {
    const snap = renderer.create(<GlobalStateComponent />).toJSON();
    expect(snap).toMatchSnapshot();
  });
});
