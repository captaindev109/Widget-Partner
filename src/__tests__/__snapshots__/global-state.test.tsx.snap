// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots Snapshot test should pass 1`] = `
Array [
  <div
    data-testid="state"
  >
    TRY_HEDVIG
  </div>,
  <button
    data-testid="forward"
    onClick={[Function]}
  >
    forward
  </button>,
  <button
    data-testid="backward"
    onClick={[Function]}
  >
    backward
  </button>,
]
`;
